### ISSUE TITLE
Batch support

### ISSUE THREAD

**Issue by danroth27** [aspnet/BasicMiddleware#246]
_From @rosieks on January 8, 2015 17:19_

ASP.NET WebAPI 2 introduced support for batching. There was issue for batch implementation in Mvc but it was closed: https://github.com/aspnet/Mvc/issues/508

Despite that it would be great to have this usefull feature.

_Copied from original issue: aspnet/Home#262_

**Comment by danroth27** [comment]
_From @Praburaj on January 8, 2015 18:20_

/CC @danroth27 @yishaigalatzer

**Comment by danroth27** [comment]
_From @damienbod on May 27, 2015 21:7_

How will this be supported? 
Greetings Damien

**Comment by danroth27** [comment]
_From @OneCyrus on May 20, 2016 11:12_

+1 
batch support is essential for modern web apps. any progress or new information?

**Comment by danroth27** [comment]
_From @rdefreitas on May 20, 2016 11:35_

the original issue suggested doing this through Middleware. It might not be too difficult to implement.

**Comment by danroth27** [comment]
_From @brianoconnell on August 25, 2016 17:32_

Has there been any further plans made to support batching either via middleware or otherwise?

**Comment by danroth27** [comment]
_From @gilmishal on June 4, 2017 11:39_

This shouldn't be too difficult, Would have done it my self if it wasn't for my lack of time... It seems like it requires a router implementation mapping `/batch` to multiple `RouteContext`s by formatting the text input (json, http format text) each with their own Handler based on the default configuration (i.e MVC).
it might be a little bit more difficult, because it would require support for multiple responses for an HttpContext (as far as I know it seems like it will be shared across Controllers, which kinda makes it not possible at the moment).

**Comment by danroth27** [comment]
_From @josh-sachs on June 14, 2017 22:56_

+1 important.

**Comment by danroth27** [comment]
_From @khushalpatel1981 on June 18, 2017 12:22_

+2 Very Important, it's stopped us to migrate from .net 4.5 to core 1.1 and
move to production

**Comment by srikrsna** [comment]
I've managed to parse a batch request, I suppose it's fairly straight forward, but what escapes me at the moment is as to how I can forward all the requests internally and batch their responses. Any idea as to how to make internal requests from middle ware and record responses?

**Comment by danroth27** [comment]
I chatted with @Tratcher and we think it would look something like this:
- Create a new DefaultHttpContext
- Mock out any features you need to
- Call next with their new context
- Serialize back the response.

**Comment by srikrsna** [comment]
@danroth27  @Tratcher Thank you for the quick reply. Seems like a solution to me but I can't really wrap my head around "Mock out any features you need", could you elaborate or point me to any sample. Thanks again.

**Comment by Tratcher** [comment]
That bit is going to be pretty in depth, you have to re-implement or copy many of the underlying features provided by the server. See https://github.com/aspnet/HttpAbstractions/tree/dev/src/Microsoft.AspNetCore.Http.Features

At a bare minimum you must implement IHttpRequestFeature and IHttpResponseFeature, or at least provide the associated data using a default implementation: https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http/Features/HttpRequestFeature.cs

**Comment by Tornhoof** [comment]
As I'm one of those who need a BatchHandler too, I made a ProofOfConcept in a few hours, which basically copies much of the code from the Testserver https://github.com/aspnet/Hosting/tree/dev/src/Microsoft.AspNetCore.TestHost

Technically it is fairly simple, at least as far as I tried, most of the TestServer code actually does pretty much the same.
Main problem is, I had to add the old System.Http.Formatting.dll back in, because it contains all the nice MultipartContent helpers, including HttpMessageContent and all the other stuff required to properly parse a Multipart batch operation and create the response
Simply said, without anyone porting all those nice helpers over to Core, this will never work and as far as I know there are no plans to port those over.

POC:
[BatchHandler.zip](https://github.com/aspnet/BasicMiddleware/files/1236589/BatchHandler.zip)

It basically just calls a few GET methods on the demo values controller.

My bottom line on this POC is simple, I can't port my WebAPI 2.0 application to ASP.NET Core 2.0 without a batch handler.  Personally I think that ASP.NET Core 2.0 is still lacking many nice "Quality-of-Life" features which were provided by ``HttpRequestMessage`` and ``HttpResponseMessage``.

**Comment by Tratcher** [comment]
@Tornhoof do you need to parse incoming multipart or generate outgoing multipart? ASP.NET Core has a few different layers of API for parsing incoming multipart. See HttpRequest.Form, MultipartReader, etc..

There isn't a multipart writer, but that part is much easier to write.

**Comment by Tornhoof** [comment]
@Tratcher 
I need to do both :)
Let's look at the steps for a second:
1. Splitting Incoming Multipart Request: ``MultiPartReader/Section`` (incl. in .ASP.NET Core)
2. Parsing the ``application/http`` multipart section (the individual batched http request): ``HttpMessageContent`` (not included in ASP.NET Core), using: https://github.com/ASP-NET-MVC/aspnetwebstack/blob/master/src/System.Net.Http.Formatting/HttpMessageContent.cs
3. Sending it of to the pipeline (taken from the TestServer project, incl. in ASP.NET Core)
4. Producing MultiPart output (not included in ASP.NET Core, using parts from the old stack for that)

The old stack gives me methods and classes to parse the section into a ``HttpRequestMessage``, a typed object with all the nice properties describing request and methods and classes to serialize a  ``HttpResponseMessage`` into a ``HttpContent`` with all the proper headers etc. in place.

That's what I meant with "Quality-of-Life" features, no ``ReadAsByteAsync``, ``ReadAsStringAsync``, no nice typed headers etc. With the old stack, it was way easier to work with HttpRequestMessage/ResponseMessage, sending it via ``HttpClient``, receiving/responding in a Controller. Now the new ``HttpRequest`` and ``HttpResponse`` are completely separated and no easy method of conversion exists (yet).

**Comment by Tratcher** [comment]
I see serialization code in HttpMessageContent, but not parsing code. Where's that?

**Comment by Tratcher** [comment]
Ah, this one? https://github.com/ASP-NET-MVC/aspnetwebstack/blob/4e40cdef9c8a8226685f95ef03b746bc8322aa92/src/System.Net.Http.Formatting/HttpContentMessageExtensions.cs#L217

**Comment by Tornhoof** [comment]
Yes, this ist the Parser method. Sorry, I forgot to mention that it is implemented in the Extension

**Comment by Tratcher** [comment]
@davidfowl it looks like your generic HTTP parsers would be useful here.

**Comment by Tornhoof** [comment]
Smaller update from my side
- I replaced the old WebAPI 2 reading/writing part in the middleware with my own code. The Tests still include the old api.

I updated the POC for that:
[BatchHandler.zip](https://github.com/aspnet/BasicMiddleware/files/1269761/BatchHandler.zip)

I'm not really sure if the complicated async offload, responsestream etc. is really necessary as this is still copied from the TestServer project. As there is no "client" and the batch is done sequentially for now, I guess it could be simplified quite a bit.

**Comment by jkotalik** [comment]
@Tornhoof would it be possible to put your PoC on Github for ease of browsing?

**Comment by Tornhoof** [comment]
@jkotalik 
Here you go: https://github.com/Tornhoof/HttpBatchHandler

**Comment by Tornhoof** [comment]
Update from my side:
The POC in https://github.com/Tornhoof/HttpBatchHandler is starting to look ok, all my tests pass (both in the handler and in my main application):
What's in there?

1. OnXXX Events are added
2. Tests are added

What's missing?

1. Error Cases (those are a bit undefined in the old webapi handler, the basic ones exist)
2. Conversion of tests not to use old webapi nugets to create multipart messages

**Comment by Tornhoof** [comment]
I'm asking myself where to put the handler as soon as it is in a good enough shape for public consumption.
I currently see four options:

1. Try to contribute to this repository, including the necessary code changes etc.
2. Try to get into WebAPIContrib.Core
3. Create my own nuget package
4. Hope that my repo is enough of a starting point from someone "official" to do work on this issue.

I guess the first one is probably preferable, as it will increase the code quality, since obviously people with vastly more experience in ASP.NET Core are present.

**Comment by Tratcher** [comment]
https://github.com/aspnet-contrib would also be an option

**Comment by Tornhoof** [comment]
Thanks for the suggestion of aspnet-contrib.

So I replaced the old webapi nugets for multipart messages with new implementations, so this should be independent of the old libraries now.

One question: Why are there no ``ConfigureAwait(false)`` in the Middlewares of ASP.NET Core?
As far as I understood it, it's best practice to use ConfigureAwait(false) in libraries and ASP.NET core has no custom context which would require ``ConfigureAwait(true)``. Just asking, because I usually add explicit ConfigureAwait to everything and I don't know what to do with the middleware.

**Comment by Tornhoof** [comment]
There is a difference I saw on a test system recently between my batch handler for Core and the one from the old ASP.NET.
Assume the following URL http://www.mysystem.com/webapp/api/mystuff
In the old handler the batched request would look like this:
```
--batch_357647d1-a6b5-4e6a-aa73-edfc88d8866e
Content-Type: application/http; msgtype=request

POST /webapp/api/mystuff HTTP/1.1
Host: mysystem
Content-Type: application/json; charset=utf-8

{"Id":129,"Name":"Name4752cbf0-e365-43c3-aa8d-1bbc8429dbf8"}
```
The host is mysystem and the uri for the Http request line includes the PathQuery part of the full uri (including **webapp**)

In the handler for core, kestrel apparently does not really know much about the **webapp** part and routing internally only works without the **webapp** part. So it looks like this
```
--batch_357647d1-a6b5-4e6a-aa73-edfc88d8866e
Content-Type: application/http; msgtype=request

POST /api/mystuff HTTP/1.1
Host: mysystem
Content-Type: application/json; charset=utf-8

{"Id":129,"Name":"Name4752cbf0-e365-43c3-aa8d-1bbc8429dbf8"}
```
Now there are two questions:

1. Is that correct behaviour?, it's different from the old one, I guess some compatibility system should be in place?
2. What about calculated urls in the response (e.g. location header), I guess the URL should be a valid external one (e.g. http://www.mysystem.com/webapp/api/mystuff)

**Comment by Tornhoof** [comment]
Hmm, looking at https://github.com/aspnet/Hosting/blob/b207658f22ca1b9322e7f141c318e9c2b5db4f3f/src/Microsoft.AspNetCore.TestHost/ClientHandler.cs#L140-L151 It's solved there by setting the pathbase appropriately

**Comment by SlimShaggy** [comment]
@Tornhoof you have lots of .ConfigureAwait(false) in your code, but I don't think it's necessary because ASP.Net Core does not have SynchronizationContext.

**Comment by Tornhoof** [comment]
@SlimShaggy  Yes, but I prefer explicit ConfigureAwaits over implicit ConfigureAwaits and I annotate them with false as much as possible. The general recommendation still is (afaik) to use ConfigureAwait(false) in Library code (where possible).
